method_comparison/peft_bench/utils.py

```diff
@@ -0,0 +1,453 @@
+            {
+                "stage": stage,
+                "ram_mb": ram_mb,
+                "gpu_allocated_mb": gpu_allocated_mb,
+                "gpu_reserved_mb": gpu_reserved_mb,
+            }
+        )
+
+    def add_metrics_for_category(self, category: str, metrics: dict, individual_samples: list = None):
+        """Add metrics for a specific prompt category under generation_info."""
+        category_data = {
+            "metrics": metrics,
+            "samples": individual_samples if individual_samples is not None else []
+        }
+        self.generation_info["by_category"][category] = category_data
+
+    def update_run_info(self, duration: float, status: BenchmarkStatus, error: Optional[str] = None, peft_config: Optional[dict] = None, benchmark_config: Optional[dict] = None):
+        """Update run information."""
+        self.run_info["duration"] = duration
+        self.run_info["status"] = status.value
+        if error:
+            self.run_info["error"] = error
+        if peft_config:
+            self.run_info["peft_config"] = peft_config
+        if benchmark_config:
+            self.run_info["benchmark_config"] = benchmark_config
+
+    def compute_overall_metrics(self):
+        """Compute overall metrics across all categories within generation_info."""
+        if not self.generation_info["by_category"]:
+            return
+
+
+        categories = self.generation_info["by_category"]
+        key_metrics = ["inference_time", "base_inference_time", "inference_overhead_pct", "time_per_token", "generated_tokens"]
+        
+        for metric in key_metrics:
+            values = []
+            for category_data in categories.values():
+                if "metrics" in category_data and metric in category_data["metrics"]:
+                    values.append(category_data["metrics"][metric])
+            
+            if values:
+                self.generation_info["overall"][metric] = sum(values) / len(values)
+
+    def to_dict(self) -> dict[str, Any]:
+        """Convert result to dictionary."""
+        self.compute_overall_metrics()
+        return {
+            "run_info": self.run_info,
+            "generation_info": self.generation_info,
+            "meta_info": self.meta_info,
+        }
+
+    def save(self, path: Optional[str] = None):
+        """Save result to JSON file."""
+        if path is None:
+            # Determine the appropriate path based on status
+            if self.status == BenchmarkStatus.SUCCESS:
+                # If successful, use the main result path
+                base_path = RESULT_PATH
+            elif self.status == BenchmarkStatus.FAILED:
+                # If failed, use the cancelled path
+                base_path = RESULT_PATH_CANCELLED
+            elif self.status == BenchmarkStatus.CANCELLED:
+                # If explicitly cancelled, use the cancelled path
+                base_path = RESULT_PATH_CANCELLED
+            else:
```

Comment:
How can we reach this branch? It's not really possible, is it? Note that in MetaMathQA, we store the results in the temp result folder iff the branch != `main`. This way, we ensure that the `results/` is reserved for results from the `main` branch. I think we should do the same here. Check the code here:

https://github.com/huggingface/peft/blob/04a5ed7b2f1d35f11c1522a686181ed08c01d7bb/method_comparison/MetaMathQA/utils.py#L655-L662